### PR TITLE
WIP: Kopia Repository Upgrade Lock

### PR DESCRIPTION
**Pull request by Shrekster**
https://docs.google.com/document/d/1zIghQc4i4vJSANjSDBOvjzMiqnZSEp1cCasxSUkltSk/edit#heading=h.khston2oxmxy

Upgrading from format 1 to 3:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner --io-drain-timeout 1s --status-poll-interval 1s --max-clock-drift 1s
Repository upgrade lock intent has been placed.
Waiting for 2s to allow all other kopia clients to drain ...
Waiting for 1s to allow all other kopia clients to drain ...
Waiting for 0s to allow all other kopia clients to drain ...
Successfully drained all repository clients, the lock has been fully-established now.
migrating current indexes to epoch format
Repository indices have been upgraded.
Repository has been successfully upgraded
```

Upgrading from format 3 to 3:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner --io-drain-timeout 1s --status-poll-interval 1s --max-clock-drift 1s
ERROR error setting the upgrade lock intent: repository is already upgraded to 3, we can only upgrade till 3
```

Upgrading from format 2 to format 3:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner --io-drain-timeout 1s --status-poll-interval 1s --max-clock-drift 1s
Repository upgrade lock intent has been placed.
repository indices have already been migrated to the epoch format
Repository has been successfully upgraded
```

Non-owner upgraders during an upgrade:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id ownerx --io-drain-timeout 10s --status-poll-interval 5s --max-clock-drift 1s
ERROR failed to open repository: repository upgrade in progress
ERROR open repository: unable to open repository: repository upgrade in progress
```

Connections to repository during an upgrade:
```sh
$ kopia repository connect filesystem --password 12345678 --config-file /tmp/kopia/repository.config --path /tmp/kopia/repo/
ERROR failed to open repository: repository upgrade in progress
deleted repository password for /tmp/kopia/repository.config.
kopia: error: error connecting to repository: repository upgrade in progress, try --help
```

Upgrade status with advance notice:
```sh
$ kopia repository status --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner
Config file:         /tmp/kopia/repository.config

Description:         Repository in Filesystem: /tmp/kopia/repo
Hostname:            c02fq0ttmd6r
Username:            shikhar.mall
Read-only:           false
Format blob cache:   15m0s

Storage type:        filesystem
Storage config:      {
                       "path": "/tmp/kopia/repo",
                       "fileMode": 384,
                       "dirMode": 448,
                       "dirShards": [
                         3,
                         3
                       ]
                     }

Unique ID:           83e7ec60180d56181c979213b22b9401c3709d1fcac1a2b2d9d8b921940e0c7e
Hash:                BLAKE2B-256-128
Encryption:          AES256-GCM-HMAC-SHA256
Splitter:            DYNAMIC-4M-BUZHASH
Format version:      3
Content compression: false
Password changes:    false
Max pack length:     20 MiB
Index Format:        v1
Epoch Manager:       disabled
Ongoing upgrade:     Format Version 3 -> 3
Upgrade Time:        2021-12-22 17:05:27.98409 -0800 PST
Upgrade lock:        Unlocked
Lock status:         Draining
```

Upgrade status with immediate lock:
```sh
$ kopia repository status --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner
Config file:         /tmp/kopia/repository.config

Description:         Repository in Filesystem: /tmp/kopia/repo
Hostname:            c02fq0ttmd6r
Username:            shikhar.mall
Read-only:           false
Format blob cache:   15m0s

Storage type:        filesystem
Storage config:      {
                       "path": "/tmp/kopia/repo",
                       "fileMode": 384,
                       "dirMode": 448,
                       "dirShards": [
                         3,
                         3
                       ]                                                                                                                                                                                                                                                                             }

Unique ID:           83e7ec60180d56181c979213b22b9401c3709d1fcac1a2b2d9d8b921940e0c7e
Hash:                BLAKE2B-256-128
Encryption:          AES256-GCM-HMAC-SHA256
Splitter:            DYNAMIC-4M-BUZHASH
Format version:      3
Content compression: false
Password changes:    false
Max pack length:     20 MiB
Index Format:        v1
Epoch Manager:       disabled
Ongoing upgrade:     Format Version 1 -> 3
Upgrade Time:        2021-12-22 16:55:49.98409 -0800 PST
Upgrade lock:        Locked
Lock status:         Fully Established
```

### CODE REVIEW

File: site/content/docs/Upgrade/_index.md [hunk 1]
```diff
@@ -40,22 +40,24 @@ $ kopia repository disconnect
...
 * make sure to stop any running `kopia server` instances and disable all background kopia tasks, such as periodic snapshots in `crontab`.
 
-3. Upgrade all kopia clients to the latest version >=v0.9.x
+3. Upgrade all kopia clients to the latest version >=v0.10.x
 
 4. Using the designated `kopia` client, run:
 
 ```
-$ kopia repository set-parameters --upgrade
+$ kopia repository upgrade --upgrade-owner-id <unique-id>
```
Comment (jkowalski) [inline site/content/docs/Upgrade/_index.md:48]: let's discuss the CLI experience on slack, perhaps?

**Review by jkowalski — COMMENTED** [review]
I took early look, not very deep yet.

Overall looks great!

File: repo/content/content_upgrade_lock.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+// the repository. The lock object is set on the Kopia repository format blob
+// 'kopia.repository' and must be respected by all clients accessing the
+// repository.
+type UpgradeLock struct {
+	OwnerID                string        `json:"ownerID,omitempty"`
+	CreationTime           time.Time     `json:"creationTime,omitempty"`
+	AdvanceNoticeDuration  time.Duration `json:"advanceNoticeDuration,omitempty"`
+	IODrainTimeout         time.Duration `json:"ioDrainTimeout,omitempty"`
+	StatusPollInterval     time.Duration `json:"statusPollInterval,omitempty"`
+	OldFormatVersion       FormatVersion `json:"oldFormatVersion,omitempty"`
```
[the remark this answers is not in the document]
Comment (Shrekster) [reply repo/content/content_upgrade_lock.go:20]: Display and also to restore the original version if someone attempts to force revoke the lock.

Although, it is debatable whether trying to restore the old format version is a good thing on force revoke.

File: repo/content/content_upgrade_lock.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		IODrainTimeout:         ioDrainTimeout,
+		StatusPollInterval:     statusPollInterval,
+		MaxPermittedClockDrift: maxPermittedClockDrift,
+		OldFormatVersion:       oldFormatVersion,
+	}
+}
+
+// Update upgrades an existing lock intent. This method controls what mutations
+// are allowed on an upgrade lock once it has been placed on the repository.
+func (l *UpgradeLock) Update(other *UpgradeLock) (*UpgradeLock, error) {
```
[the remark this answers is not in the document]
Comment (Shrekster) [reply repo/content/content_upgrade_lock.go:44]: Makes sense. Will do.

File: repo/open.go [hunk 4]
```diff
@@ -224,12 +272,21 @@ func openWithConfig(ctx context.Context, st blob.Storage, lc *LocalConfig, passw
...
+	}
+
 	cmOpts := &content.ManagerOptions{
 		RepositoryFormatBytes: fb,
 		TimeNow:               defaultTime(options.TimeNowFunc),
 		DisableInternalLog:    options.DisableInternalLog,
 	}
 
+	if locked, _ := repoConfig.UpgradeLock.IsLocked(cmOpts.TimeNow()); locked && options.UpgradeOwnerID != repoConfig.UpgradeLock.OwnerID {
+		return nil, ErrRepositoryUnavailableDueToUpgrageInProgress
```
[the remark this answers is not in the document]
Comment (Shrekster) [reply repo/open.go:287]: That is a good point. Also, there are certain implications of doing it that way. Let's discuss this offline.

File: site/content/docs/Upgrade/_index.md (hunk 1 above)
Comment (Shrekster) [reply site/content/docs/Upgrade/_index.md:48]: UPDATE from offline talk:

We decided to maintain backwards compatibility so we do not break existing usage and also start moving to a dedicated upgrade CLI. I will also update docs around this.

**Comment by Shrekster** [comment]
Resolved conflicts past the retention blob related merges.

**Pull request closed without merging**